### PR TITLE
Disable ec2 mirror for non aws instances

### PR DESCRIPTION
**Pull request by lucasmoura**
For versions before 20.2, we allowed the use of ec2 mirrors if the datasource availability_zone matches one of the ec2 regions. We are now updating that behavior to allow allow the use of ec2 mirrors on ec2 instances or if the user directly passes an an ec2 mirror url through #cloud-config apt directives. 

LP: #1456277

### CODE REVIEW

File: cloudinit/distros/__init__.py [hunk 1]
```diff
@@ -845,7 +848,13 @@ def _get_package_mirror_info(mirror_info, data_source=None,
         # ec2 availability zones are named cc-direction-[0-9][a-d] (us-east-1b)
         # the region is us-east-1. so region = az[0:-1]
         if _EC2_AZ_RE.match(data_source.availability_zone):
-            subst['ec2_region'] = "%s" % data_source.availability_zone[0:-1]
+            ec2_region = data_source.availability_zone[0:-1]
+
+            if (ALLOW_EC2_MIRRORS_ON_NON_AWS_INSTANCE_TYPES
+                    and data_source.platform_type == "ec2"):
+                subst['ec2_region'] = "%s" % ec2_region
+            elif not ALLOW_EC2_MIRRORS_ON_NON_AWS_INSTANCE_TYPES:
```
Comment (blackboxsw) [inline cloudinit/distros/__init__.py:856]: This elif is incorrect isn't it? We only want to set subst[ec2_region'] if any([ data_source.platform_type == "ec2", ALLOW_EC2_MIRRORS_ON_NON_AWS_INSTANCE_TYPES] ?

File: cloudinit/distros/__init__.py [hunk 2]
```diff
@@ -845,7 +848,13 @@ def _get_package_mirror_info(mirror_info, data_source=None,
         # ec2 availability zones are named cc-direction-[0-9][a-d] (us-east-1b)
         # the region is us-east-1. so region = az[0:-1]
         if _EC2_AZ_RE.match(data_source.availability_zone):
```
Comment (blackboxsw) [inline cloudinit/distros/__init__.py:850]: I think maybe we can just drop this re.match and instead just specificall check data_source.platform_type == "ec2".

File: cloudinit/distros/__init__.py (hunk 1 above)
Comment (lucasmoura) [reply cloudinit/distros/__init__.py:856]: Yes, it is incorrect :facepalm: I will update the logic to fix this

File: cloudinit/distros/__init__.py (hunk 2 above)
Comment (lucasmoura) [reply cloudinit/distros/__init__.py:850]: But wouldn't that interfere with past release ? We would like to have past releases still being able to user ec2 mirrors, right ?

**Comment by lucasmoura** [comment]
@blackboxsw I have update the PR with the suggested changes. There is just some things that I have some doubts about:

1) I have verified that when we use a jinja template file as you stated, we get the parsed version on the apt_configure module when we start evaluating the mirrors. But where do you think we should put an example of this change ? Maybe we can add it as an example on the apt_configure module schema ?
2) I know this is not related to this PR, but just to make sure, we we call this [method](https://github.com/canonical/cloud-init/blob/1933a4a2ce6c0903da46cd2bbc407936503f96a1/cloudinit/stages.py#L544) do we parse the user-data and apply it at the same time or am I missing something ?
3) I have not yet addressed the distro url in this update, but I do think this is a good idea. Shall I add this to this PR or another one ?

File: cloudinit/distros/__init__.py (hunk 2 above)
Comment (blackboxsw) [reply cloudinit/distros/__init__.py:854]: You are right, darn backward compatibility.

File: cloudinit/features.py [hunk 3]
```diff
@@ -26,6 +26,16 @@
...
 """
 
+
+ALLOW_EC2_MIRRORS_ON_NON_AWS_INSTANCE_TYPES = False
+"""
+When configuring apt mirrors, old behavior is to allow
+the use of ec2 mirrors if the user region matches one
+of the possible aws regions. We are now updating that to only
+the use of ec2 mirror for aws instances, unless the user specifically
+sets that on the userdata config.
```
Comment (blackboxsw) [inline cloudinit/features.py:36]:
```suggestion
the use of ec2 mirrors if the datasource availability_zone format
matches one of the possible aws ec2 regions. After the 20.2 release, we
no longer publish ec2 region mirror urls on non-AWS cloud platforms.
Besides feature_overrides.py, users can override this by providing
#cloud-config apt directives.
```

**Comment by lucasmoura** [comment]
@blackboxsw Done, I have updated both descriptions

**Pull request merged**